overlord/snapstate/handlers.go

```diff
@@ -950,13 +950,164 @@ func asyncRefreshOnSnapClose(st *state.State, refreshInfo *userclient.PendingSna
 			continueInhibitedAutoRefresh(st)
 		}
 	}()
 
 	return nil
 }
 
+// asyncBeginDeferredRefreshNotification broadcasts a desktop notification  in a goroutine,
+// telling the user that a refresh process is being started, and that they must wait before
+// trying to launch the snap.
+//
+// This allows the, possibly slow, communication with each snapd session agent,
+// to be performed without holding the snap state lock.
+var asyncBeginDeferredRefreshNotification = func(context context.Context, client *userclient.Client, finishInfo *userclient.FinishedSnapRefreshInfo) {
+	go func() {
+		if err := client.BeginDeferredRefreshNotification(context, finishInfo); err != nil {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+// asyncFinishDeferredRefreshNotification broadcasts desktop notification in a goroutine,
+// telling the user that a refresh process has been completed and that they can now launch
+// the snap.
+//
+// This allows the, possibly slow, communication with each snapd session agent,
+// to be performed without holding the snap state lock.
+var asyncFinishDeferredRefreshNotification = func(context context.Context, client *userclient.Client, finishInfo *userclient.FinishedSnapRefreshInfo) {
+	go func() {
+		if err := client.FinishDeferredRefreshNotification(context, finishInfo); err != nil {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+// asyncSnapRefreshProgressNotification broadcasts desktop notification in a goroutine,
+// to update the current refresh status.
+//
+// This allows the, possibly slow, communication with each snapd session agent,
+// to be performed without holding the snap state lock.
+var asyncSnapRefreshProgressNotification = func(context context.Context, client *userclient.Client, progressInfo *userclient.SnapRefreshProgressInfo) {
+	go func() {
+		if err := client.SnapRefreshProgressNotification(context, progressInfo); err != nil {
+			logger.Noticef("Cannot send notification about finishing deferred refresh: %v", err)
+		}
+	}()
+}
+
+// calculateTaskProgress search the task list and count how many tasks remain
+// to be completed for the snap that is being refreshed, to be able to show in
+// the popup dialog how many tasks have been already been done.
+// Also, it takes the name of the task being done right now to show it in the
+// dialog.
+//
+// It must be called with st locked. It isn't locked/unlocked inside to allow
+// to call it from asyncRefreshOnSnapClose(), where it is already locked.
+func calculateTaskProgress(st *state.State, snapName string, instanceName string, revision snap.Revision) (int, int, string) {
+	remainingTasks := 0
+	totalTasks := 0
+	newBarText := ""
+	for _, task := range st.Tasks() {
+		snapsup, _, err := snapSetupAndState(task)
+		if err != nil {
+			continue
+		}
+		if snapsup.SnapName() != snapName {
```

Comment:
The point is... "instanceName" is always unique in all the system?

Reply:
I ask this because, according to the documentation, "instanceName" is for parallel installations, and the whole name is composed by "snapName" plus "instanceName"...

Reply:
Fixed.